### ISSUE TITLE
Text file with repeated content failed to be downloaded

### ISSUE THREAD

**Issue by islamTaha12** [zero-os/0-stor#225]
```bash
➜  zerostorcli git:(master) ✗ zerostorcli file upload /home/upload/upload_file_106                    
file uploaded, key = upload_file_106
➜  zerostorcli git:(master) ✗ zerostorcli file download upload_file_106 /home/upload/upload_file_106_1
download file failed: snappy: corrupt input
➜  zerostorcli git:(master) ✗ 
```

**Config:**
```bash
organization: xtremx90 # itsyou.online organization name
namespace: namespace_test       # itsyou.online namespace
protocol: rest
iyo_app_id:
iyo_app_secret: 
# the address(es) of 0-stor server
data_shards:
    - http://10.147.17.234:8080
    - http://10.147.17.66:8080
    - http://10.147.17.205:8080
    - http://10.147.17.85:8080
# the address(es) of etcd server
meta_shards:
    - http://10.147.17.234:2379

block_size: 4096

replication_nr: 4
replication_max_size: 4096

distribution_data: 3
distribution_parity: 1

compress: true
encrypt: true
encrypt_key: ab345678901234567890123456789012
```

**File size**
-rw-r--r-- 1 root root 89817 Aug 30 15:44 /home/upload/upload_file_106

**Affected version**
https://github.com/zero-os/0-stor/commit/a19fec7a8ec562c59134fd869e483f523f118844

**Comment by zaibon** [comment]
Turns out the problem comes from the file that @islamTaha12 used to do his test.

For some reason the way he generated the file creates files that can't be downloaded.

**Comment by zaibon** [comment]
Here is the file that is impossible to download
[test_file.zip](https://github.com/zero-os/0-stor/files/1285179/test_file.zip)

**Comment by iwanbk** [comment]
i suspect that the chunker doesn't do the streaming properly

**Comment by zaibon** [comment]
> i suspect that the chunker doesn't do the streaming properly

What makes you think that ?
Cause I tried to upload a file generated from /dev/uradom of the exact same size and it worked.

**Comment by iwanbk** [comment]
I got similar issue, not sure if it is the same thing.

When i wrote unit test for new zeroctl copy, which copy tlog data from one vdisk to another, the code do the `Walk` twice.
First walk is OK, but the second walk got same error as this one
```
snappy: corrupt input 
```

**Comment by zaibon** [comment]
Hum this is more serious then I thought then.
@iwanbk do you have the dataset that also make walk fails ?

**Comment by iwanbk** [comment]
Nope, it is just random data generated by `rand.Read`.
I'm checking it now

**Comment by iwanbk** [comment]
i can confirm that it is my own fault, there is nothing wrong in 0-stor side

**Comment by GlenDC** [comment]
@zaibon, @iwanbk, @islamTaha12 any more updates on this issue? this is marked as `critical`, yet reported 2 months ago... This is really impossible for a critical issue. Either close it as invalid, or fix it ASAP, but this can't just remain open for as long as it is.

**Comment by iwanbk** [comment]
@chrisvdg IIRC you could reproduce it, please add your steps to reproduce it

**Comment by chrisvdg** [comment]
I could not as mine was corrupted with concurrent writing to the same key, in normal conditions my file was always downloadable.

**Comment by GlenDC** [comment]
@iwanbk can you add a system test for this test scenario, based on the steps to reproduce from above. If that test passes, we can just close this issue for now.

**Comment by GlenDC** [comment]
I added bunch of unit tests, using (or inspired by) the linked file in https://github.com/zero-os/0-stor/issues/225#issuecomment-327831767, and it seems that the bug has something to with the combination of chunked content and erasure coding. However if I test the erasure coder seperately, with async chunked content, it does all work out.

Therefore it must have something to do with how we store the chunked+erasure_coded data, because it's only when we combine chunking+erasure_coding+storage that this bug gets triggered in my unit test.

(For example chunking+storage on its own does work, same for chunking+replication+storage.)

But the bug is still there I think. The good thing is that there are now unit tests for my on machine, so triggering the bug is as easy as hitting a single button, just need to fix it now. Just that... 🌵

**Comment by GlenDC** [comment]
Unexpectedly, the origins of this bug seems to be related to #421.

The problem is that we currently write all different parts of the erasure coding using the same key, which obviously means that we're overwriting parts, and thus corrupts the data, and because of that we are not able to read our files back.

This also makes it clear that why we only suffer this specific bug when using erasure-code distribution, as even though the content might be the same, there are different parts, and as we choose the shards (on which we store these parts) randomly, it means that we write different parts to different shards, overwriting perhaps other kinds of parts of other data. So when you do that for a file full of repeated content, in a chunked manner, it means that you have all your blocks (chunks) sharing the same key. Knowing that it's not hard to see why this bug triggers 100% of the time using a large enough file, with just 1 repeated character.

To my fix-225 branch I added for example a test to test the pipeline against some other files (the root README of this project and the root CHANGELOG of this project), to get a sense of how those same tests do against a file with more realistic data, rather than all repeated. And you can see, the tests do pass for those files. This because in such real-world files, the different chunks will be actually always different (as even 1 byte difference gives a different hash), and thus the chance of collision is not there (it exists still but it is too low to happen for the little data we write). And as the distribution write logic guarantees that it only ever uses any given random shard once, we do not get collisions, and thus we can read our file back just fine.

Therefore in order to fix this issue, we need to fix issue #421.

This makes this bug yet another example of how flawed our key-generation scheme currently is. The way we currently generate keys is really very brute, overwriting content in all kind of places.

Talking about a plot twist...

**Comment by zaibon** [comment]
The fact that we use  the same key for each part is already a bug. 
But even if we change that, it won't solve #421.
